Makefile

```diff
@@ -1,52 +1,386 @@
+	$(call BUILD,darwin,amd64,-darwin-amd64)
+bin/git-lfs-darwin-386 : $(SOURCES)
+	$(call BUILD,darwin,386,-darwin-386)
+bin/git-lfs-linux-amd64 : $(SOURCES)
+	$(call BUILD,linux,amd64,-linux-amd64)
+bin/git-lfs-linux-386 : $(SOURCES)
+	$(call BUILD,linux,386,-linux-386)
+bin/git-lfs-freebsd-amd64 : $(SOURCES)
+	$(call BUILD,freebsd,amd64,-freebsd-amd64)
+bin/git-lfs-freebsd-386 : $(SOURCES)
+	$(call BUILD,freebsd,386,-freebsd-386)
+bin/git-lfs-windows-amd64.exe : resource.syso $(SOURCES)
+	$(call BUILD,windows,amd64,-windows-amd64.exe)
+bin/git-lfs-windows-386.exe : resource.syso $(SOURCES)
+	$(call BUILD,windows,386,-windows-386.exe)
+
+# .DEFAULT_GOAL sets the operating system-appropriate Git LFS binary as the
+# default output of 'make'.
+.DEFAULT_GOAL := bin/git-lfs$(X)
+
+# bin/git-lfs targets the default output of Git LFS on non-Windows operating
+# systems, and respects the build knobs as above.
+bin/git-lfs : $(SOURCES) fmt
+	$(call BUILD,$(GOOS),$(GOARCH),)
 
-all : bin/git-lfs
+# bin/git-lfs targets the default output of Git LFS on Windows systems, and
+# respects the build knobs as above.
+bin/git-lfs.exe : $(SOURCES) resource.syso
+	$(call BUILD,$(GOOS),$(GOARCH),.exe)
 
-include $(MAKEFILE_GEN)
+# resource.syso install the 'goversioninfo' command and uses it in order to
+# generate a binary that has information included necessary to create the
+# Windows installer.
+resource.syso:
+	@$(GO) get github.com/josephspurrier/goversioninfo/cmd/goversioninfo
+	$(GO) generate
 
-$(LIBDIR)/git-lfs.o : $(SRC_main) $(DEPS_main)
-	@mkdir -p $(dir $@)
-	$(GOC) $(GOFLAGS) -c -o $@ $(SRC_main)
+# RELEASE_TARGETS is the set of all release artifacts that we generate over a
+# particular release. They each have a corresponding entry in BUILD_TARGETS as
+# above.
+#
+# Unlike BUILD_TARGETS above, each of the below create a compressed directory
+# containing the matching binary, as well as the contents of RELEASE_INCLUDES
+# below.
+RELEASE_TARGETS = bin/releases/git-lfs-darwin-amd64-$(VERSION).tar.gz \
+	bin/releases/git-lfs-darwin-386-$(VERSION).tar.gz \
+	bin/releases/git-lfs-linux-amd64-$(VERSION).tar.gz \
+	bin/releases/git-lfs-linux-386-$(VERSION).tar.gz \
+	bin/releases/git-lfs-freebsd-amd64-$(VERSION).tar.gz \
+	bin/releases/git-lfs-freebsd-386-$(VERSION).tar.gz \
+	bin/releases/git-lfs-windows-amd64-$(VERSION).zip \
+	bin/releases/git-lfs-windows-386-$(VERSION).zip
 
-bin/git-lfs : $(LIBDIR)/git-lfs.o $(DEPS_main)
-	@mkdir -p $(dir $@)
-	$(GOC) $(GOFLAGS) -o $@ $^
+# RELEASE_INCLUDES are the names of additional files that are added to each
+# release artifact.
+RELEASE_INCLUDES = README.md CHANGELOG.md
 
-%.a : %.o
-	$(AR) rc $@ $<
+# release is a phony target that builds all of the release artifacts, and then
+# shows the SHA 256 signature of each.
+.PHONY : release
+release : $(RELEASE_TARGETS)
+	shasum -a 256 $(RELEASE_TARGETS)
 
+# bin/releases/git-lfs-%-$(VERSION).tar.gz generates a gzip-compressed TAR of
```

Comment:
It might be worth putting an example here, like `make bin/releases/git-lfs-darwin-amd64-v2.4.0-354-g27ff38c6.tar.gz`. I that might make it a bit clearer how to use this make target

Reply:
Good call, I added some documentation to this effect in edfcac99 (Makefile: add some extra documentation for building release artifacts, 2018-07-25). I also added a note on how to build Git LFS versions with a custom VERSION name, which you can do as described in that commit.